Added a hook for implementing UoW on top of SynchronizedStorageSession

SzymonPobiega opened the pull request:
Adds `RegisterCommitHook` method to the storage session. It allows registering custom unit of work completion logic that is executed before committing the storage session transaction.

andreasohlund commented on src/NServiceBus.NHibernate/SynchronizedStorage/INHibernateSynchronizedStorageSession.cs:
```diff
@@ -1,9 +1,13 @@
     {
         ISession Session { get; }
+
+        void RegisterCommitHook(Func<Task> callback);
```
Would there be value in trying to align with NH event listener lingo

http://nhibernate.info/doc/nhibernate-reference/events.html

In that case perhaps: `OnCommit(...)` ?

SzymonPobiega replied:
@andreasohlund I don't think we should aim for alignment as the goal would be to have the same hook in native SQL persistence.  On the other hand, I am not attached to the current name so if you thing `OnCommit` is better, I am 👍 to change.

andreasohlund replied:
Wouldn't we want  to call this `.OnSaveChanges(...)` in raven?

SzymonPobiega commented:
@Particular/nhibernate-persistence-maintainers please take a look. I think this is ready.

SzymonPobiega commented:
Ping :-) cc @tmasternak

MarcinHoppe commented on src/NServiceBus.NHibernate/CallbackList.cs:
```diff
@@ -0,0 +1,48 @@
+        Func<Task> singleItem;
+        List<Func<Task>> multipleItems;
+
+        public void Add(Func<Task> callback)
```
Is the effect of storing single item separately measurable? It would be good if this complexity was justified by a significant benefit.

MarcinHoppe commented on src/NServiceBus.NHibernate/CallbackList.cs:
```diff
@@ -0,0 +1,48 @@
+            {
+                foreach (var callback in multipleItems)
+                {
+                    await callback().ConfigureAwait(false);
```
Is there a reason to run these callbacks asynchronously but still sequentially. We could start them to run in parallel and await `Task.WhenAll(...)`.

SzymonPobiega replied:
Hmmm... Maybe going with multiple callbacks was a bit over-engineered and we should stick with non-additive `set`-style API? Then there would be no problems with concurrency and the list of callbacks would not be necessary.

MarcinHoppe replied:
❤️ 

Let's release a simple thing and get feedback.

andreasohlund commented:
You had me at "simple" :heart:

On Mon, Mar 27, 2017 at 11:52 AM, Marcin Hoppe <notifications@github.com>
wrote:

> *@MarcinHoppe* commented on this pull request.
> ------------------------------
>
> In src/NServiceBus.NHibernate/CallbackList.cs
> <https://github.com/Particular/NServiceBus.NHibernate/pull/266#discussion_r108130565>
> :
>
> > +            {
> +                singleItem = callback;
> +            }
> +        }
> +
> +        public async Task InvokeAll()
> +        {
> +            if (singleItem != null)
> +            {
> +                await singleItem().ConfigureAwait(false);
> +            }
> +            else if (multipleItems != null)
> +            {
> +                foreach (var callback in multipleItems)
> +                {
> +                    await callback().ConfigureAwait(false);
>
> ❤️
>
> Let's release a simple thing and get feedback.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Particular/NServiceBus.NHibernate/pull/266#discussion_r108130565>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHoZN0t5sXMSEH1N6gZOxVFyfU6NTsPks5rp4bfgaJpZM4MHSSJ>
> .
>

tmasternak commented on src/NServiceBus.NHibernate/SynchronizedStorage/NHibernateLazyNativeTransactionSynchronizedStorageSession.cs:
```diff
@@ -22,23 +23,28 @@ public NHibernateLazyNativeTransactionSynchronizedStorageSession(Func<ISession>
-        public Task CompleteAsync()
+        public async Task CompleteAsync()
         {
+            await callbacks.InvokeAll().ConfigureAwait(false);
```
What if callbacks throw?

SzymonPobiega commented:
I've simplified the API by removing the possibility for registering multiple callbacks and added unit tests to prove that session is not committed when callback throws an exception.

MarcinHoppe commented on src/NServiceBus.NHibernate/Outbox/NHibernateOutboxTransaction.cs:
```diff
@@ -25,17 +25,23 @@ public void Dispose()
         {
-            await callbacks.InvokeAll().ConfigureAwait(false);
-
+            if (onSaveChangesCallback != null)
```
I think you can get rid of this `if` statement using the elvis operator (`?.`) and the `Invoke` method.

SzymonPobiega replied:
Wouldn't this return `null` as a result and then `await` would blow up?

MarcinHoppe replied:
You are right, I forgot about the async part. Ignore my previous comment.

MarcinHoppe commented on src/NServiceBus.NHibernate/SynchronizedStorage/NHibernateAmbientTransactionSynchronizedStorageSession.cs:
```diff
@@ -36,7 +40,10 @@ public void Dispose()
         public async Task CompleteAsync()
         {
-            await callbacks.InvokeAll().ConfigureAwait(false);
+            if (onSaveChangesCallback != null)
```
Ditto.

MarcinHoppe commented on src/NServiceBus.NHibernate/SynchronizedStorage/NHibernateLazyNativeTransactionSynchronizedStorageSession.cs:
```diff
@@ -23,16 +23,23 @@ public NHibernateLazyNativeTransactionSynchronizedStorageSession(Func<ISession>
         public async Task CompleteAsync()
         {
-            await callbacks.InvokeAll().ConfigureAwait(false);
+            if (onSaveChangesCallback != null)
```
Same here.

SzymonPobiega commented:
@tmasternak can you give it another try?

tmasternak commented on src/NServiceBus.NHibernate/Outbox/NHibernateOutboxTransaction.cs:
```diff
@@ -22,12 +23,25 @@ public void Dispose()
+
+        public async Task Commit()
+        {
+            if (onSaveChangesCallback != null)
```
Do we need to call `Transaction.Dispose` even if callback throws?

SzymonPobiega replied:
Added disposing off of the transaction to `Dispose`

tmasternak commented on src/NServiceBus.NHibernate/SynchronizedStorage/NHibernateAmbientTransactionSynchronizedStorageSession.cs:
```diff
@@ -20,29 +21,34 @@ public NHibernateLazyAmbientTransactionSynchronizedStorageSession(Func<IDbConnec
+            {
+                await onSaveChangesCallback(this).ConfigureAwait(false);
+            }
+            if (session.IsValueCreated)
```
Same question as above about `session.Value.Dispose()`

SzymonPobiega replied:
I think this is fine as the session is used in a `using` block and the dispose will take care of `session.Value`.

tmasternak commented on src/NServiceBus.NHibernate/SynchronizedStorage/NHibernateLazyNativeTransactionSynchronizedStorageSession.cs:
```diff
@@ -9,10 +9,11 @@
         public NHibernateLazyNativeTransactionSynchronizedStorageSession(Func<ISession> sessionFactory)
         {
-            session =new Lazy<ISession>(() =>
+            session = new Lazy<ISession>(() =>
```
Do we need `s` declaration in labmda? Maybe just `return ...`

SzymonPobiega replied:
Now that I store the result of `BeginTransaction` it can't be done ;-)

tmasternak commented on src/NServiceBus.NHibernate/SynchronizedStorage/NHibernateLazyNativeTransactionSynchronizedStorageSession.cs:
```diff
@@ -22,23 +23,35 @@ public NHibernateLazyNativeTransactionSynchronizedStorageSession(Func<ISession>
+            if (onSaveChangesCallback != null)
+            {
+                await onSaveChangesCallback(this).ConfigureAwait(false);
+            }
```
Same question about `Dispose`

SzymonPobiega replied:
Added disposing off of the transaction to `Dispose`

SzymonPobiega commented:
@tmasternak I've fixed what I could ;-)
